### PR TITLE
Safer definition lists implementation

### LINKED ISSUE #973: pulldown-cmark panics on first pass when enabling `definition-lists`
I'm not sure if nested definition lists are valid, but using [pulldown-cmark 0.12.2](https://github.com/pulldown-cmark/pulldown-cmark/tree/v0.12.2) and the following input causes a panic:

```markdown
level one!
:
    level two!
    :
        level three!
        :

new level one!
:
```

Here's a small program that reproduces the issue:
```rust
use pulldown_cmark::{Options, Parser};

static INPUT: &str = "
level one!
:
    level two!
    :
        level three!
        :

new level one!
:
";

fn main() {
    let options = Options::ENABLE_DEFINITION_LIST;
    let parser = Parser::new_ext(INPUT, options);
    for _ in parser {}
}

```

<details>
<summary>Traceback</summary>

```
thread 'main' panicked at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/firstpass.rs:1540:38:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/panicking.rs:74:14
   2: core::panicking::panic
             at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/panicking.rs:148:5
   3: core::option::unwrap_failed
             at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/option.rs:2012:5
   4: core::option::Option<T>::unwrap
             at /Users/ytmimi/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/option.rs:972:21
   5: pulldown_cmark::firstpass::FirstPass::pop
             at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/firstpass.rs:1540:22
   6: pulldown_cmark::firstpass::FirstPass::parse_block
             at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/firstpass.rs:99:13
   7: pulldown_cmark::firstpass::FirstPass::run
             at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/firstpass.rs:76:18
   8: pulldown_cmark::firstpass::run_first_pass
             at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/firstpass.rs:41:5
   9: pulldown_cmark::parse::Parser<F>::new_with_broken_link_callback
             at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/parse.rs:270:34
  10: pulldown_cmark::parse::Parser::new_ext
             at /Users/ytmimi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/parse.rs:255:9
  11: example::main
             at ./src/example.rs:19:18
  12: core::ops::function::FnOnce::call_once
             at /Users/ytmimi/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
```

</details>

### PR DESCRIPTION
**Pull request by mondeja**
Fixes #973

### CODE REVIEW

**Comment by Martin1887** [comment]
What do you think, @notriddle?

**Comment by notriddle** [comment]
As near as I can tell, this doesn't break anything. :+1:

File: pulldown-cmark/src/tree.rs [hunk 1]
```diff
@@ -186,8 +186,6 @@ impl<T: Default> Tree<T> {
         }
         if next.is_some() {
             self.cur = next;
-        } else {
-            self.pop();
```
Comment (notriddle) [inline pulldown-cmark/src/tree.rs:190]: I have no idea what I was thinking when I wrote this. It doesn't make sense, and it doesn't seem necessary for anything.

Comment (mondeja) [reply pulldown-cmark/src/tree.rs:190]: I though the same finding the bug. Looks to me like some sort of premature optimization.

**Comment by Martin1887** [comment]
OK, thanks, then this can be merged after fixing the CI build!

**Comment by notriddle** [comment]
@mondeja I think you can just [rebase this branch](https://rustc-dev-guide.rust-lang.org/git.html#rebasing) and the CI problem will go away. A fix for it has already been merged into the master branch.

    git pull --rebase https://github.com/pulldown-cmark/pulldown-cmark.git master
    git push -f

**Pull request merged**